packages/adblocker/src/engine/engine.ts

```diff
@@ -1443,29 +1447,97 @@ export default class FilterEngine extends EventEmitter<EngineEventHandlers> {
       let redirectRule: NetworkFilter | undefined;
 
+      // 1st priority: @@||<entity>$removeparam
+      // 2nd priority: ||<entity>$removeparam
+      // 3rd priority: @@||<entity>$removeparam=x
+      // 4th priority: ||<entity>$removeparam=x
+      let removeparamExceptionFilter: NetworkFilter | undefined;
+      let redirectUrl: URL | undefined;
+
       // If `result.filter` is `undefined`, it means there was no $important
-      // filter found so far. We look for a $redirect filter.  There is some
-      // extra logic to handle special cases like redirect-rule and
-      // redirect=none.
-      //
-      // * If redirect=none is found, then cancel all redirects.
-      // * Else if redirect-rule is found, only redirect if request would be blocked.
-      // * Else if redirect is found, redirect.
+      // filter found so far. We look for $removeparam and $redirect filter.
       if (result.filter === undefined) {
-        const redirects = this.redirects
-          .matchAll(request, this.isFilterExcluded.bind(this))
-          // highest priorty wins
-          .sort((a, b) => b.getRedirectPriority() - a.getRedirectPriority());
+        const redirectableFilters = this.redirects.matchAll(
+          request,
+          this.isFilterExcluded.bind(this),
+        );
+        const redirectFilters: NetworkFilter[] = [];
+        const removeparamFilters: Map<string, NetworkFilter> = new Map();
+        const removeparamExceptionFilters: Map<string, NetworkFilter> = new Map();
+        for (const filter of redirectableFilters) {
+          if (filter.isRemoveParam()) {
+            if (filter.isException()) {
+              removeparamExceptionFilters.set(filter.removeparam!, filter);
+            } else {
+              removeparamFilters.set(filter.removeparam!, filter);
+            }
+          } else {
+            redirectFilters.push(filter);
+          }
+        }
+
+        // We don't need to match `removeparam` at all if:
+        // * `removeparam` filters not matched
+        // * URL doesn't have search parameter
+        if (removeparamFilters.size !== 0) {
+          redirectUrl = new URL(request.url);
+          if (redirectUrl.searchParams.size !== 0) {
+            for (const [key, filter] of removeparamFilters) {
+              // Remove all params in case of option value is empty
+              // We will not match individual exceptions since it has a higher priority than them
+              if (key === '') {
+                result.filter = filter;
+                removeparamExceptionFilter = removeparamExceptionFilters.get('');
+
+                // In case of non-existence of global exception, we will remove params
+                if (removeparamExceptionFilter === undefined) {
+                  redirectUrl.search = '';
+                }
+
+                break;
+              }
+
+              if (redirectUrl.searchParams.has(key)) {
+                result.filter = filter;
```

Comment:
That seems invalid in case there is more than one `removeparam` filter applying to the URL. I also think that in terms of semantics, the `result.filter` means a blocking filter, but removeparam filters aren't blocking requests. Maybe would it make more sense to add an additional key on `result` with a list of removeparams (or if we want to generalize, a list of URL/request modifiers)

Reply:
Outdated. Shape changed.